sw/airborne/modules/ctrl/optical_flow_landing.c

```diff
@@ -112,56 +133,80 @@ PRINT_CONFIG_VAR(OPTICAL_FLOW_LANDING_OPTICAL_FLOW_ID)
+    int16_t flow_y, int16_t flow_der_x, int16_t flow_der_y, float quality, float size_divergence, float dist);
 
+// struct containing most relevant parameters
 struct OpticalFlowLanding of_landing_ctrl;
 
 void vertical_ctrl_module_init(void);
 void vertical_ctrl_module_run(bool in_flight);
 
+// for exponential gain landing, gain increase per second during the first (hover) phase:
+#define INCREASE_GAIN_PER_SECOND 0.02
+
 /**
  * Initialize the optical flow landing module
  */
 void vertical_ctrl_module_init(void)
 {
   unsigned int i;
 
+  // filling the of_landing_ctrl struct with default values:
   of_landing_ctrl.agl = 0.0f;
   of_landing_ctrl.agl_lp = 0.0f;
   of_landing_ctrl.vel = 0.0f;
-  of_landing_ctrl.divergence_setpoint = 0.0f;
-  of_landing_ctrl.cov_set_point = -0.025f;
-  of_landing_ctrl.cov_limit = 0.0010f; //1.0f; // for cov(uz,div)
-  of_landing_ctrl.lp_factor = 0.95f;
+  of_landing_ctrl.divergence_setpoint = 0.0f; //-0.20f; // For exponential gain landing, pick a negative value
+  of_landing_ctrl.cov_set_point = -0.0075f; // for cov(uz, div), i.e., cov_method 0
+  of_landing_ctrl.cov_limit =
+    2.2f; // This high a value means that for a constant divergence landing no landing will be triggered
+  // If you want to trigger a landing, set the limit to something like 0.025f; // limit for cov(uz,div) - used only for landing triggering
+  of_landing_ctrl.lp_factor = 0.75f; // for Bebop 2  // 0.60f; // for AR drone
   of_landing_ctrl.pgain = OPTICAL_FLOW_LANDING_PGAIN;
   of_landing_ctrl.igain = OPTICAL_FLOW_LANDING_IGAIN;
   of_landing_ctrl.dgain = OPTICAL_FLOW_LANDING_DGAIN;
   of_landing_ctrl.sum_err = 0.0f;
-  of_landing_ctrl.nominal_thrust = 0.710f; //0.666f; // 0.640 with small battery
+  of_landing_ctrl.d_err = 0.0f;
+  of_landing_ctrl.nominal_thrust = 0.630f; // 0.710f; //0.666f; // 0.640 with small battery
   of_landing_ctrl.VISION_METHOD = OPTICAL_FLOW_LANDING_VISION_METHOD;
   of_landing_ctrl.CONTROL_METHOD = OPTICAL_FLOW_LANDING_CONTROL_METHOD;
   of_landing_ctrl.COV_METHOD = OPTICAL_FLOW_LANDING_COV_METHOD;
-  of_landing_ctrl.delay_steps = 40;
+  of_landing_ctrl.delay_steps = 15;
+  of_landing_ctrl.window_size = OPTICAL_FLOW_LANDING_COV_WINDOW_SIZE;
   of_landing_ctrl.pgain_adaptive = 10.0;
   of_landing_ctrl.igain_adaptive = 0.25;
   of_landing_ctrl.dgain_adaptive = 0.00;
+  of_landing_ctrl.reduction_factor_elc =
+    0.80f; // for exponential gain landing, after detecting oscillations, the gain is multiplied with this factor
+  of_landing_ctrl.lp_cov_div_factor =
+    0.99; // low pass filtering cov div so that the drone is really oscillating when triggering the descent
+  of_landing_ctrl.count_transition =
+    300; // tuned for Bebop 2 (higher frame rate than AR drone) - number of time steps the low-passed cov div should be beyond the limit
+  of_landing_ctrl.p_land_threshold =
+    0.15f; // if the gain reaches this value during an exponential landing, the drone makes the final landing.
 
   struct timespec spec;
```

Comment:
You should use the paparazzi time functions to make time platform independent. I can implement if you like.